File: solutions/python/sliding-window/subarray-product-less-than-k.py
# LeetCode 713 — Subarray Product Less Than K (Medium)
# Category: Sliding Window · Approach: Sliding Window
# Time: O(n) | Space: O(1)
# Source: https://leetcode.com/problems/subarray-product-less-than-k/

def num_subarray_product_less_than_k(nums: list[int], k: int) -> int:
    if k <= 1:
        return 0
    prod = 1
    left = 0
    count = 0
    for right, x in enumerate(nums):
        prod *= x
        while prod >= k:
            prod //= nums[left]
            left += 1
        count += right - left + 1
    return count

File: solutions/typescript/sliding-window/subarray-product-less-than-k.ts
// LeetCode 713 — Subarray Product Less Than K (Medium)
// Category: Sliding Window · Approach: Sliding Window
// Time: O(n) | Space: O(1)
// Source: https://leetcode.com/problems/subarray-product-less-than-k/

function numSubarrayProductLessThanK(nums: number[], k: number): number {
  if (k <= 1) return 0;
  let prod = 1;
  let left = 0;
  let count = 0;
  for (let right = 0; right < nums.length; right++) {
    prod *= nums[right];
    while (prod >= k) {
      prod /= nums[left];
      left++;
    }
    count += right - left + 1;
  }
  return count;
}
